designs/extensible-cli-and-scaffolding-plugins.md

```diff
@@ -0,0 +1,144 @@
+# Extensible CLI and Scaffolding Plugins
+
+## Overview 
+I would like for Kubebuilder to become more extensible, such that it could be imported and used as a library in other projects. Specifically, I'm looking for a way to use Kubebuilder's existing CLI and scaffolding for Go projects, but to also be able to augment Kubebuilder's project versions with other custom project versions so that I can support the Kubebuilder workflow with non-Go operators (e.g. operator-sdk's Ansible and Helm-based operators).
+
+The idea is for Kubebuilder to define one or more plugin interfaces that can be used to drive what the `init`, `create api` and `create webhooks` subcommands do and to add a new `cli` package that other projects can use to integrate out-of-tree plugins with the Kubebuilder CLI in their own projects.
+
+## Related issues and PRs
+
+* [#1148](https://github.com/kubernetes-sigs/kubebuilder/pull/1148)
+* [#1171](https://github.com/kubernetes-sigs/kubebuilder/pull/1171)
+* Possibly [#1218](https://github.com/kubernetes-sigs/kubebuilder/issues/1218)
+
+## Prototype implementation
+https://github.com/joelanford/kubebuilder-exp
+
+## Plugin interfaces
+
+### Required
+Each plugin would minimally be required to implement the `Plugin` interface.
+
+```go
+type Plugin interface {
+    // Version is the project version that this plugin implements.
+    // For example, Kubebuilder's Go v2 plugin implementation would return "2"
+    Version() string
+}
+```
+
+### Optional
+Next, a plugin could optionally implement further interfaces to declare its support for specific Kubebuilder subcommands. For example:
+* `InitPlugin` - to initialize new projects
+* `CreateAPIPlugin` - to create APIs (and possibly controllers) for existing projects
+* `CreateWebhookPlugin` - to create webhooks for existing projects
+
+Each of these interfaces would follow the same pattern (see the InitPlugin interface example below).
+
+```go
+type InitPlugin interface {
+    Plugin
+
+    // InitDescription returns a description of what this plugin initializes
+    // for a new project. It is used to display help.
+    InitDescription() string
+
+    // InitHelp returns one or more examples of the command-line usage
+    // of this plugin's project initialization support. It is used to display help.
+    InitExample() string
+
+    // BindInitFlags binds the plugin's init flags to the CLI. This allows each
+    // plugin to define its own command line flags for the `kubebuilder init`
+    // subcommand.
+    BindInitFlags(fs *pflag.FlagSet)
+
+    // Init initializes a project.
+    Init() error
+}
```

Comment:
One high level pattern is -- There will be metadata associated with the project (ex. domain, groups...) and type (`helm`, `golang`...), we need a way to persist that data in  probably `PROJECT` file and each subcommand invocation can be provided with that metadata and can be written back at the end during the invocation or at the end.
If we can come up with a generic metadata interface (Generic Key-Value may be ?) on PROJECT file, then we can decouple commandline args from the plugins.

Reply:
Yes this sounds like a necessity. I like @DirectXMan12's [suggestion](https://github.com/kubernetes-sigs/kubebuilder/pull/1250#issuecomment-564259130) from above. Is this along the lines of what you're looking for?